### PR TITLE
Fix SoftDeleteRepositories reconcile issue due to OptimisticLockException

### PR DESCRIPTION
**Pull request by anandjakhaniya**
[Kibana](https://kibana.dev.verta.ai/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'2021-06-20T17:30:00.000Z',to:now))&_a=(columns:!(message),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:f4c6aa40-32b2-11eb-b0d6-b3168b0b5e7b,key:kubernetes.namespace_name,negate:!f,params:(query:production),type:phrase),query:(match_phrase:(kubernetes.namespace_name:production))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:f4c6aa40-32b2-11eb-b0d6-b3168b0b5e7b,key:kubernetes.labels.app,negate:!f,params:(query:modeldb--backend),type:phrase),query:(match_phrase:(kubernetes.labels.app:modeldb--backend))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:f4c6aa40-32b2-11eb-b0d6-b3168b0b5e7b,key:level,negate:!f,params:(query:ERROR),type:phrase),query:(match_phrase:(level:ERROR)))),index:f4c6aa40-32b2-11eb-b0d6-b3168b0b5e7b,interval:auto,query:(language:kuery,query:''),sort:!(!('@ts',asc))))

All DeleteEntitiesCron issues like 
```
DeleteEntitiesCron : deleteRepositories : Exception: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1;
```
in above kibana is due to OptimisticLockException which has fixed in this PR.

### CODE REVIEW

**Review by conradoverta — COMMENTED** [review]
I think I'm missing something here. I don't believe `reconcile` can be called with similar ids in parallel. So `processingIdSet` should always be empty. Do you have evidence of it not being empty?

**Comment by anandjakhaniya** [comment]
> I think I'm missing something here. I don't believe `reconcile` can be called with similar ids in parallel. So `processingIdSet` should always be empty. Do you have evidence of it not being empty?

Yes, i have reproduce it. lets assume ids being empty but now a days UAC will take time to process `deleteEntityResourcesWithServiceUser` so still those entities are not deleted and those are in process but at that time reconcile utils again wake up at schedule time and  it found those entity not deleted because it was on process from first wake up. So after going further, both reconcile jobs try to delete objects and get Optimistic Lock Exception in second one.

**Comment by conradoverta** [comment]
> > I think I'm missing something here. I don't believe `reconcile` can be called with similar ids in parallel. So `processingIdSet` should always be empty. Do you have evidence of it not being empty?
> 
> Yes, i have reproduce it. lets assume ids being empty but now a days UAC will take time to process `deleteEntityResourcesWithServiceUser` so still those entities are not deleted and those are in process but at that time reconcile utils again wake up at schedule time and it found those entity not deleted because it was on process from first wake up. So after going further, both reconcile jobs try to delete objects and get Optimistic Lock Exception in second one.

I'd rather fix this in the reconciler tool itself not to allow duplicated ids to begin with. It has the same effect, but it fixes this issue for all reconcilers at once.

**Comment by anandjakhaniya** [comment]
> I'd rather fix this in the reconciler tool itself not to allow duplicated ids to begin with. It has the same effect, but it fixes this issue for all reconcilers at once.

Ok, that would be great, i have try to implement it in `Reconciler<T>` class but i thought you will not agree with it so i do not have added it there. :)

**Review by conradoverta — COMMENTED** [review]
As discussed in [slack](https://vertaai.slack.com/archives/CJG0NL86B/p1624515605256600?thread_ts=1624515567.256400&cid=CJG0NL86B), this has to be optional.

**Comment by anandjakhaniya** [comment]
> As discussed in [slack](https://vertaai.slack.com/archives/CJG0NL86B/p1624515605256600?thread_ts=1624515567.256400&cid=CJG0NL86B), this has to be optional.

Updated PR with deduplicate logic

File: backend/common/src/main/java/ai/verta/modeldb/common/reconcilers/Reconciler.java [hunk 1]
```diff
@@ -15,23 +16,28 @@
...
+import java.util.stream.Collectors;
 
 public abstract class Reconciler<T> {
   final Logger logger;
   protected final HashSet<T> elements = new HashSet<>();
   protected final LinkedList<T> order = new LinkedList<>();
   final Lock lock = new ReentrantLock();
   final Condition notEmpty = lock.newCondition();
+  // To prevent OptimisticLockException
+  private final Set<T> processingIdSet = ConcurrentHashMap.newKeySet();
```
Comment (conradoverta) [inline backend/common/src/main/java/ai/verta/modeldb/common/reconcilers/Reconciler.java:28]: As per the the conversation in slack, this is not enough to make things thread safe.

**Review by conradoverta — CHANGES_REQUESTED** [review]

File: backend/common/src/main/java/ai/verta/modeldb/common/reconcilers/Reconciler.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/common/src/main/java/ai/verta/modeldb/common/reconcilers/Reconciler.java:28]: `ConcurrentHashMap.newKeySet();` will create the thread safe `Set`  so it is enough as per my thought.

Comment (conradoverta) [reply backend/common/src/main/java/ai/verta/modeldb/common/reconcilers/Reconciler.java:28]: It is not. Please refer to the slack thread where I described the 2 race conditions. This fixes one (as I described in the thread) but not the other

**Review by conradoverta — CHANGES_REQUESTED** [review]
Still has a race condition

**Comment by anandjakhaniya** [comment]
> Still has a race condition

Added lock for second case.

**Review by conradoverta — COMMENTED** [review]
Following up in slack for speed: https://vertaai.slack.com/archives/CJG0NL86B/p1625648035303300

**Review by conradoverta — APPROVED** [review]

**Pull request merged**